src/Lecture2.hs

```diff
@@ -58,7 +64,11 @@ lazyProduct = error "TODO"
 "ccaabb"
 -}
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
+duplicate l = dup l []
+    where
+        dup :: [a] -> [a] -> [a]
+        dup [] acc      = acc
+        dup (x:xs) acc  = dup xs (acc ++ [x, x])
```

Comment:
It is more efficient to construct lists using the cons operator (`:`) directly without introducing intermediate values or adding lists with `(++)` whenever possible. Adding elements to the end of the list is rather slow 🐌

Reply:
Ok, I got it. So I can use the cons operator and at the end reverse the list, right?

Reply:
It's better to avoid reversing the list as well. You can use the cons operator to prepend an element to the recursive call of the function itself. This way, you'll avoid both `reverse` and adding to the end

Reply:
Ohhhh I see, something like:
```
duplicate :: [a] -> [a]
duplicate [] = []
duplicate (x:xs)  = (x:x:duplicate xs)
```

Reply:
Exactly! 💯